### PR TITLE
Changed Memory Registration Cache comparison func

### PR DESCRIPTION
**Pull request by jswaro**
Changed the comparison function to search for exact matches.
This is the first PR of two to implement an optimized search
algorithm.

Signed-off-by: James Swaro jswaro@cray.com

### CODE REVIEW

**Comment by jswaro** [comment]
@sungeunchoi @hppritcha @ztiffany 

This is the first PR of two.

File: prov/gni/src/gnix_mr.c [hunk 1]
```diff
@@ -440,6 +464,10 @@ int gnix_mr_reg(struct fid *fid, const void *buf, size_t len,
 		}
 	}
 
+	reg_addr = ((uint64_t) buf) & ~((1 << GNIX_MR_PAGE_SHIFT) - 1);
+	reg_len = __calculate_length((uint64_t) buf, len,
+			1 << GNIX_MR_PAGE_SHIFT);
```
Comment (hppritcha) [inline prov/gni/src/gnix_mr.c:469]: i thought the rounding up/down was going to take in to account the pagesize info stored in the gni memhndl,

**Comment by sungeunchoi** [comment]
LGTM.  Can you open an issue to make MR thread-safe?  I can't unsee race conditions.

File: prov/gni/src/gnix_mr.c (hunk 1 above)
Comment (jswaro) [inline prov/gni/src/gnix_mr.c:469]: If the pagesize info in the gni memhandle is used, then we can't resize
the registration prior to calling GNI_MemRegister.

Additionally, there is no way to tell from the mem handle how long the
registration is. I've attempted to read the number of pages covered by the
memory handle and the number is bogus. It claims to have 353452 pages, all
4k size, or ~1.4GB for a 1k registration. That seems off to me.
-- Jim

**Comment by jswaro** [comment]
The memory registration cache is locked during registration and
deregistration. Race conditions aren't going to be possible until those
locks are removed. 
-- Jim

File: prov/gni/src/gnix_mr.c (hunk 1 above)
Comment (ztiffany) [inline prov/gni/src/gnix_mr.c:469]: We found that the length is bogus when using MDD sharing.  You should still be able to read the pagesize out of the MDH then round to find the registration length.

**Comment by ztiffany** [comment]
It's hard to review without more description of the chosen solution.  A lot was talked about and I didn't follow to the end.

**Comment by jswaro** [comment]
This revision of the memory registration cache does the following:
- changes the memory registration cache to a exact match tree based system
  In the previous system, if memory ranged from 0x1000 to 0x2000 was already registered, then a new registration from 0x1200 to 0x2000 would reuse the registration 0x1000:0x2000. In this new system, the old registration would not be reused since it doesn't have the same base address. 

This only a temporary set-back because only the red-black tree is being used, since the alternate slow path structure is not implemented. This will create more registrations with less reuse, but we will not run into the case where registrations cannot be found in the tree due to rebalance.
- removes the bug where registrations might be deallocated unintentionally by deregistering other memory registrations. 

The next PR will address reducing memory registrations by searching the tree by base address alone, and matching against the best fit for the registration, if one exists. In the prior example, the new registration, though it was a subset of an already registered region, would not have been matched. The slowpath of next PR will allow that registration to match the prior registration. 

Additionally since this current PR will not allow a new registration of 0x1000 to 0x1800 to match the prior registration of 0x1000:0x2000, that will be addressed in the next PR by the following:

All entries in the RB tree will be addresed by base address alone. Entries of the redblack tree will be a list of registrations sharing the same base addressed sorted in descending order of registration length. This will allow the fast path to quickly locate a matching request and determine whether the request can be satisfied using an existing request. Since there may not be a matching request with the same base address, the slow path will search existing registrations to see if there is an existing registration which encompasses the registration request. This is slower process since we need to organize the registrations into a list by overlapping regions.

This is being done in two parts, first to address the bugs and the second to optimize.

**Comment by ztiffany** [comment]
:+1: 

We should still look into detecting the correct registration page size in the next commit.

**Comment by hppritcha** [comment]
:+1:

**Pull request merged**